File: algo_expert/easy/nth_fibonacci/nth_fibonacci.js
const assert = require('assert');

('use strict');

function getNthFib(n) {
	// Declare the output variable to be returned.
	let nthFibNum = 0;

	// Initialize an array, with values 0 and 1 (respectively), to store
	// the Fibonacci sequence.
	const fibArr = [0, 1];

	if (n <= 0) {
		return 'Sorry. Please pass an input value greater than 0.';
	} else if (n === 1) {
		return 0;
	} else if (n === 2) {
		return 1;
	} else {
		const startIndex = 2;
		for (let i = startIndex; i < n; i++) {
			const twoNumsPrev = fibArr[i - 2];
			const oneNumPrev = fibArr[i - 1];
			const newFibNum = twoNumsPrev + oneNumPrev;
			// console.log(`fibArr_before = ${fibArr}`);
			fibArr.push(newFibNum);
			// console.log(`fibArr_after = ${fibArr}`);
		}
		nthFibNum = fibArr[n - 1];
		return nthFibNum;
	}
}

console.log('------------------TEST CASES------------------');
assert.deepStrictEqual(
	getNthFib(0),
	'Sorry. Please pass an input value greater than 0.',
	'There is no such thing as a "0th Fibonacci number."'
);
assert.deepStrictEqual(getNthFib(2), 1, '1 is the 2nd Fibonacci number');
assert.deepStrictEqual(getNthFib(4), 2, '2 is the 4th Fibonacci number');
assert.deepStrictEqual(getNthFib(6), 5, '5 is the 6th Fibonacci number');
console.log('----------ALL TEST CASES ARE PASSING----------');

// console.log(getNthFib(0)); // should return "Sorry. Please pass an input value greater than 0."
// console.log(getNthFib(2)); // should return 1
// console.log(getNthFib(4)); // should return 2
// console.log(getNthFib(6)); // should return 5

File: algo_expert/easy/nth_fibonacci/nth_fibonacci.py
def getNthFib(n):
    # Declare the output variable to be returned.
    nthFibNum = 0
    
    # Initialize the list variable, with 0 and 1 (representing the 1st & 2nd
    # Fibonacci numbers in the sequence), for storing the fibonacci sequence.
    fibList = [0, 1]
    
    # Check to see if 'n' == 0 (Oth Fibonacci Number doesn't exist)
    if n == 0:
        return f"Sorry. Please pass an input value greater than 0."
    # Check to see if 'n' == 1 (1st Fibonacci Number is 0)
    elif n == 1:
        return f"{fibList[0]}\n"
    # Check to see if 'n' == 2 (2nd Fibonacci Number is 1)
    elif n == 2:
        return f"{fibList[1]}\n"
    # Check to see if 'n' >= 3
    else:
        # 3rd FibNum = 2nd Index of 'fibList'
        # 4th FibNum = 3rd Index of 'fibList'
        startFibIndex = 2
        for i in range(startFibIndex, n):
            prev2ndNum = fibList[i - 2]
            print(f"prev2ndNum = {prev2ndNum}")
            
            prev1stNum = fibList[i - 1]
            print(f"prev1stNum = {prev1stNum}")
            
            nextFibNum = prev2ndNum + prev1stNum
            print(f"nextFibNum = {nextFibNum}")
            
            print("fibList_before =", fibList)
            fibList.append(nextFibNum)
            print("fibList_after =", fibList)
            
        # Return the last index of the 'fibList' (nth Fibonacci number)
        nthFibNum = fibList[-1]
        return f"nthFibNum = {nthFibNum}\n"

print(getNthFib(2))  # should print 1
print(getNthFib(4))  # should print 2
print(getNthFib(6))  # should print 5


